RFD-0095. Application access Okta integration. (#18129)
gravitational/teleport: rfd/0095-application-access-okta-integration.md

## What

Allow teleport users to request access to specific applications and groups and access
applications in Okta.

## Why

Today, Teleport supports [single sign-on with Okta](https://goteleport.com/docs/access-controls/sso/okta/).
Okta is a popular IdP with our customers, so improving the integration between Okta and
Teleport will be useful.

## Details

### UX

#### Application Access UI

* After authenticating into Teleport and accessing the "Applications" item in the menu,
  a list of apps sourced from Okta will appear.
* When users click on an Okta sourced app in the UI, the application will open in a separate tab.
* These applications will *not* be behind Teleport's proxy, and will redirect to the proper Okta
  sourced URL for the application. Users will be taken through Okta's authentication process for
  this process. If users are already logged into Okta, this will be a transparent passthrough to
  the application.
* If users do not have access to an application, the link will be greyed out and an option will
  be available in the drop down menu to the right that allows users to request access to the
  application.

#### Teleport CLI

> * Okta applications will show up when doing a `tsh apps ls` from the command line.

Reviewer: ```suggestion
* Okta applications will show up when doing a `tsh apps ls` from the command line, with a type "okta".
```

Reviewer: Also, what about Okta groups?

* Logging in will **not** work for Okta apps.

#### Access requests

* When a user requests access to an application or group, the existing approval workflow
  will be used to add this access to the user.

> * Changes will be reflected in the identity provider backend when this occurs.

Reviewer: Nit: I'd just say "Okta" instead of generic "identity provider".
```suggestion
* Changes will be reflected in Okta when this occurs.
```

### High level architecture

The general architecture can be seen below, which is put into more detail later in this document.

```mermaid
flowchart LR
  subgraph Teleport
    direction BT
    DB[(Teleport Database)]
    App[Application Service]
    Requests[Access Requests Service]

    subgraph OktaSvc[Okta Service]
      BG[Background Synchronization Service]
    end

    Auth[Auth Service]
    
    OktaSvc<-->DB
    App<-->DB
    App<-->Requests
    Requests<-->DB
  end

  Okta[Okta API]
  OktaSvc<-->Okta
```

### Okta service with generic identity provider interface

> An Okta service should be introduced that synchronizes Okta applications, users, and

Reviewer: As you're describing the new `okta_service`, I would give the yaml config example first ("This is how Teleport administrator will configure Okta service") instead of jumping into implementation details.

be available in the Teleport config YAML, along with associated environment variables:

| Name | Environment Variable | Required | Description |
|------|----------------------|----------|-------------|
| `api_url` | `TELEPORT_OKTA_API_URL` | :heavy_check_mark: | The API URL so that Teleport knows which Okta endpoint to hit.
| `role_prefix` | `TELEPORT_OKTA_ROLE_PREFIX` | :x: | A role prefix to prepend to each group role created in Teleport. Defaults to `okta`.
| `role_template` | `TELEPORT_OKTA_ROLE_TEMPLATE` | :x: | A role template for Teleport roles created by the Okta service. Defaults to allowing application read access for applications that belong to a specific Okta group.

The Okta API token must be configured using an environment variable or command line
argument to prevent the possibility of users accidentally checking in this key.

**Note**: The details of the `role_prefix` and `role_template` will be discussed further in [the Okta to Teleport mapping section](#okta-to-teleport-mappings)

#### YAML example

```yaml
okta_service:
  enabled: true
  api_url: https://my-okta-endpoint.okta.com
  role_prefix: okta # This is the default role prefix.
```

> #### CLI exmples (API token)

Reviewer: ```suggestion
#### CLI examples (API token)
```

```bash
# Command line argument example
teleport start --okta-api-token 12345678

# Environment variable example
TELEPORT_OKTA_API_TOKEN="12345678" teleport start
```

### Okta user traits

Okta will optionally use an `okta_user_id` trait to map Teleport users to Okta users if necessary.

```yaml
kind: user
version: v2
metadata:
  name: alice
spec:
  roles: ['devs']
  traits:
    logins: ['admin']
    kubernetes_groups: ['edit']
    okta_user_id: ['name@mydomain.com']
```

If Okta is used for logging into Teleport or the Teleport user's username is the same as the
e-mail used for Okta, this will be unnecessary.

### Background synchronization

> The background synchronization process, which synchronizes all applications from Okta, is expected

Reviewer: Mention explicitly that Okta does not support watch API.

to run roughly every 2 minutes. This process wil translate Okta users, groups, and applications
into their Teleport equivalents.

```mermaid
sequenceDiagram
  participant OktaSvc as Okta Service
  participant Okta as Okta API
  participant AppSvc as Application Access Service

  loop Every 2 minutes
    OktaSvc->>Okta: Get list of applications
    OktaSvc->>AppSvc: Synchronize applications in Teleport

    OktaSvc->>Okta: Get list of groups

    OktaSvc->>AppSvc: Synchronize groups in Teleport
  end
```

#### Okta to Teleport mappings

Okta groups and applications will be mapped by the background synchronization into Teleport roles and
applications respectively.

##### Groups

> A Teleport role will be created for each Okta group. Each group will be prepended with a configurable

Reviewer: Didn't we want to create a new Teleport resource to represent Okta groups? What's the benefit of creating a role per group?

Author: The benefit I envisioned here was being able to use Teleport's RBAC natively to determine visibility, but I believe you've got another suggestion w/r/t this later in the doc.

`/` characters.  For example, if an Okta group called `Developers` was synchronized to Teleport,
it would be called `okta/group/Developers`.

These roles use a configurable `role_template` when created which defaults to allowing users access to
applications that belong to this group. The resulting role for `okta/group/Developers` would look like:

```yaml
kind: role
version: v5
metadata:
  name: okta/group/Developers
spec:
  allow:
    app_labels:
      group: "okta/group/Developers"
```

The corresponding application labels will be populated by application synchronization.

##### Applications

When applications are synchronized with Teleport, they will be created in application access as
HTTP apps that use the `appLinks` from Okta as their URI. If there is more than one `appLink`
associated with an Okta application, it will be split into multiple applications for
each `appLink` with the unique name of each `appLink` used to disambiguate them. The `origin` field
in the application metadata will be set to `okta`.

Applications in Okta can be assigned to groups, and the groups that the application is assigned
to will be used to create labels for the application.

> One complicating factor for us is that Okta allows per-application user assignment as well. As a

Reviewer: Why don't we extend our Role resource to support access control to specific apps/groups instead of creating roles and using app labels? For example:

```yaml
kind: role
version: v5
metadata:
  name: example
spec:
  allow:
    okta_apps: ["app1", "app2"]
    okta_groups: ["group1", "group2"]
```

Author: This makes sense. I'll adjust!

result, we'll need to create a role for each application with a list of users assigned to it and allow
it to access the application which will have a label based off of the application's unique ID
within Okta. The label and role will look like `okta/app/123456789`. Unfortunately, Okta allows
duplicate application names, so we'll have to use the actual unique identifier here. The role will have
the same role template that are used by groups, but with the application specific label targeted.

Using the same prefixes listed in the previous section for groups, if an application `Slack` is
assigned to `Developers` and `Sales`, the resulting application object would look like:

```yaml
kind: app
version: v3
metadata:
  name: Slack

>   origin: okta

Reviewer: I think the label name is `teleport.dev/origin`. https://github.com/gravitational/teleport/blob/master/api/types/constants.go#L330

### Security

* It's possible to capture sensitive data in our internal Teleport database. This could be
  undesirable for customers or other users of Teleport. We should take care that the sensitive
  identity provider data (like Okta API tokens) are never communicated to or stored outside of
  the configuration file.
* The mapping of Okta users to the appropriate Teleport user is a concern. If a user has been
  mismapped to an Okta user, it will be possible for that user to request access on behalf of
  that user. Administrators should take care with their Okta identity mappings, or even better,
  ensure that Teleport logins map directly to Okta users.

### Implementation plan

#### Okta service configuration

Implement the ability to configure the Okta service. Doing this first will make subsequent
testing and verification easier as we'll be able to pass in API URLs and API tokens.
Part of this will include implementing any stubs needed for the Okta service itself.

#### Okta service API communication

The Okta service will be able to communicate with Okta and retrieve lists of
applications and groups without doing anything with them yet.

> ####

Reviewer: Unfinished?

Author: Oh huh. I filled all this out, but I guess it didn't save? Sorry about that.
